packages/eslint-plugin/src/rules/no-unsafe-enum-assignment.ts

```diff
@@ -0,0 +1,336 @@
+    const typeChecker = parserServices.program.getTypeChecker();
+
+    /**
+     * Similar to `getEnumTypes`, but returns early as soon as it finds one.
+     */
+    function hasEnumType(type: ts.Type): boolean {
+      return tsutils
+        .unionTypeParts(type)
+        .some(subType => util.isTypeFlagSet(subType, ts.TypeFlags.EnumLiteral));
+    }
+
+    function getTypeFromNode(node: TSESTree.Node): ts.Type {
+      return typeChecker.getTypeAtLocation(
+        parserServices.esTreeNodeToTSNodeMap.get(node),
+      );
+    }
+
+    /**
+     * @returns Whether the recipient type is an enum and the provided type
+     * unsafely provides it a number.
+     */
+    function isProvidedTypeUnsafe(
+      providedType: ts.Type,
+      recipientType: ts.Type,
+    ): boolean {
+      // This short-circuits most logic: if the types are the same, we're happy.
+      if (providedType === recipientType) {
+        return false;
+      }
+
+      // `any` types can't be reasoned with
+      if (
+        tsutils.isTypeFlagSet(recipientType, ts.TypeFlags.Any) ||
+        tsutils.isTypeFlagSet(providedType, ts.TypeFlags.Any)
+      ) {
+        return false;
+      }
+
+      // If the two types are containers, check each matching type recursively.
+      //
+      // ```ts
+      // declare let fruits: Fruit[];
+      // fruits = [0, 1];
+      // ```
+      if (
+        util.isTypeReferenceType(recipientType) &&
+        util.isTypeReferenceType(providedType)
+      ) {
+        return isProvidedReferenceValueMismatched(providedType, recipientType);
+      }
+
+      // If the recipient is not an enum, we don't care about it.
+      const recipientEnumTypes = new Set(
+        getEnumTypes(typeChecker, recipientType),
+      );
+      if (recipientEnumTypes.size === 0) {
+        return false;
+      }
+
+      const providedUnionTypes = tsutils.unionTypeParts(providedType);
+
+      // Either every provided type should match the recipient enum...
+      if (
+        providedUnionTypes.every(providedType =>
+          recipientEnumTypes.has(getBaseEnumType(typeChecker, providedType)),
+        )
+      ) {
+        return false;
+      }
+
+      // ...or none of them can be an enum at all
+      return !providedUnionTypes.every(tsutils.isEnumType);
+    }
+
+    /**
+     * Finds the first mismatched type reference: meaning, a non-enum type in
+     * the provided type compared to an enum type in the recipient type.
+     */
+    function isProvidedReferenceValueMismatched(
+      providedType: ts.TypeReference,
+      recipientType: ts.TypeReference,
+    ): boolean {
+      const providedTypeArguments = typeChecker.getTypeArguments(providedType);
+      const recipientTypeArguments =
+        typeChecker.getTypeArguments(recipientType);
+
+      for (
+        let i = 0;
+        i <
+        Math.min(recipientTypeArguments.length, providedTypeArguments.length);
+        i += 1
+      ) {
```

Comment:
this formatting makes me physically ill, prettier

Reply:
whenever prettier does this it means you are doing too much on a single line.
the math.min should be on a separate line with a good variable name

Reply:
Wow, both semicolons of the for loop looks like a normal line ending semicolons. :-)

Caching that `Math.min` would be even faster because both length are constant AFAIU.

Reply:
That's a good point. The loop condition is evaluated each iteration.

Granted the loop should only be a couple of iterations, but regardless should probs extract it outwards.